src/ceph-volume/ceph_volume/devices/raw/list.py

```diff
@@ -69,7 +69,7 @@ def __init__(self, argv):
     def generate(self, devs=None):
         logger.debug('Listing block devices via lsblk...')
         info_devices = disk.lsblk_all(abspath=True)
-        if devs is None or devs == []:
+        if devs is None or devs == [] or devs == [None]:
```

Comment:
@guits The crash I encountered is identical to the crash in the original bug report. I'm pretty sure I now understand why it happened: a device had been removed from the system and other devices renamed. The rook activate script is intended to catch this situation. In its first invocation it calls `ceph-volume raw list <previous device name which no longer exists>'. This previously returned no output, but the regression in 17.2.7 means that it now crashes instead. This also prevents the activate script from continuing to scan all devices and discover the new device name.

I don't understand how this would fix that. `devs` exists, is not empty, and its contents are not `None`. The problem is that the device is not listed in info_devices because it no longer exists.

My PR https://github.com/ceph/ceph/pull/54400 fixes this, btw.

Reply:
> @guits The crash I encountered is identical to the crash in the original bug report. I'm pretty sure I now understand why it happened: a device had been removed from the system and other devices renamed. The rook activate script is intended to catch this situation. In its first invocation it calls `ceph-volume raw list <previous device name which no longer exists>`. This previously returned no output, but the regression in 17.2.7 means that it now crashes instead. This also prevents the activate script from continuing to scan all devices and discover the new device name.

interesting information that I didn't have yesterday, sorry if I have missed that.

> I don't understand how this would fix that. `devs` exists, is not empty, and its contents are not `None`. The problem is that the device is not listed in info_devices because it no longer exists.

```console
[root@cv1-1 /]# pvcreate /dev/vdb
  Physical volume "/dev/vdb" successfully created.
[root@cv1-1 /]# vgcreate vg_test /dev/vdb
  Volume group "vg_test" successfully created
[root@cv1-1 /]# lvcreate -n lv1 -L 100G vg_test
  Logical volume "lv1" created.
[root@cv1-1 /]# ceph-volume raw prepare --bluestore --data /dev/mapper/vg_test-lv1
Running command: /usr/bin/ceph-authtool --gen-print-key
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new 13f61ade-b5c9-4594-90ba-a869ff4137f1
Running command: /usr/bin/ceph-authtool --gen-print-key
Running command: /usr/bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-0
Running command: /usr/bin/chown -h ceph:ceph /dev/mapper/vg_test-lv1
Running command: /usr/bin/chown -R ceph:ceph /dev/dm-0
Running command: /usr/bin/ln -s /dev/mapper/vg_test-lv1 /var/lib/ceph/osd/ceph-0/block
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring mon getmap -o /var/lib/ceph/osd/ceph-0/activate.monmap
 stderr: got monmap epoch 1
--> Creating keyring file for osd.0
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0/keyring
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0/
Running command: /usr/bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 0 --monmap /var/lib/ceph/osd/ceph-0/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-0/ --osd-uuid 13f61ade-b5c9-4594-90ba-a869ff4137f1 --setuser ceph --setgroup ceph
 stderr: 2023-11-08T15:46:39.535+0000 7f5c24f8c640 -1 bluestore(/var/lib/ceph/osd/ceph-0//block) _read_bdev_label unable to decode label /var/lib/ceph/osd/ceph-0//block at offset 102: void bluestore_bdev_label_t::decode(ceph::buffer::v15_2_0::list::const_iterator&) decode past end of struct encoding: Malformed input [buffer:3]
 stderr: 2023-11-08T15:46:39.536+0000 7f5c24f8c640 -1 bluestore(/var/lib/ceph/osd/ceph-0//block) _read_bdev_label unable to decode label /var/lib/ceph/osd/ceph-0//block at offset 102: void bluestore_bdev_label_t::decode(ceph::buffer::v15_2_0::list::const_iterator&) decode past end of struct encoding: Malformed input [buffer:3]
 stderr: 2023-11-08T15:46:39.536+0000 7f5c24f8c640 -1 bluestore(/var/lib/ceph/osd/ceph-0//block) _read_bdev_label unable to decode label /var/lib/ceph/osd/ceph-0//block at offset 102: void bluestore_bdev_label_t::decode(ceph::buffer::v15_2_0::list::const_iterator&) decode past end of struct encoding: Malformed input [buffer:3]
 stderr: 2023-11-08T15:46:39.536+0000 7f5c24f8c640 -1 bluestore(/var/lib/ceph/osd/ceph-0/) _read_fsid unparsable uuid
--> ceph-volume raw clear prepare successful for: /dev/mapper/vg_test-lv1
[root@cv1-1 /]# CEPH_VOLUME_DEBUG=1 ceph-volume raw activate --osd-id 0 --osd-uuid 13f61ade-b5c9-4594-90ba-a869ff4137f1 --no-systemd
Traceback (most recent call last):
  File "/usr/sbin/ceph-volume", line 11, in <module>
    load_entry_point('ceph-volume==1.0.0', 'console_scripts', 'ceph-volume')()
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 41, in __init__
    self.main(self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 59, in newfunc
    return f(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 153, in main
    terminal.dispatch(self.mapper, subcommand_args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/main.py", line 32, in main
    terminal.dispatch(self.mapper, self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/activate.py", line 166, in main
    systemd=not self.args.no_systemd)
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
    return func(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/activate.py", line 75, in activate
    found = direct_report(devs)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/list.py", line 21, in direct_report
    return _list.generate(devices)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/list.py", line 91, in generate
    info_device = [info for info in info_devices if info['NAME'] == dev][0]
IndexError: list index out of range
```
after applying my patch:

```
[root@cv1-1 /]# CEPH_VOLUME_DEBUG=1 ceph-volume raw activate --osd-id 0 --osd-uuid 13f61ade-b5c9-4594-90ba-a869ff4137f1 --no-systemd
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
Running command: /usr/bin/ceph-bluestore-tool prime-osd-dir --path /var/lib/ceph/osd/ceph-0 --no-mon-config --dev /dev/mapper/vg_test-lv1
Running command: /usr/bin/chown -h ceph:ceph /dev/mapper/vg_test-lv1
Running command: /usr/bin/chown -R ceph:ceph /dev/dm-0
Running command: /usr/bin/ln -s /dev/mapper/vg_test-lv1 /var/lib/ceph/osd/ceph-0/block
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
--> ceph-volume raw activate successful for osd ID: 0
[root@cv1-1 /]#
```

> My PR #54400 fixes this, btw.

Applying your patch only didn't fix the failure I reproduced. That being said, you are right, passing a non-existing device makes ceph-volume fail the same way. so after some tests I've just made, I'm realizing there are actually 2 issues here, we might need a combination of the two patches.
I think it was better to make code change suggestions instead of opening another PR...

Reply:
@mdbooth cherry-picked your commit here

Reply:
Thanks!

Reply:
@mdbooth just tested the version of the branch, it fixes well both issues:

```console
[root@cv1-1 /]# CEPH_VOLUME_DEBUG=1 ceph-volume raw activate --osd-id 0 --osd-uuid 3680e588-0939-473c-a6d8-7d428f9b6d38 --no-systemd
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
Running command: /usr/bin/ceph-bluestore-tool prime-osd-dir --path /var/lib/ceph/osd/ceph-0 --no-mon-config --dev /dev/mapper/vg_test-lv1
Running command: /usr/bin/chown -h ceph:ceph /dev/mapper/vg_test-lv1
Running command: /usr/bin/chown -R ceph:ceph /dev/dm-0
Running command: /usr/bin/ln -s /dev/mapper/vg_test-lv1 /var/lib/ceph/osd/ceph-0/block
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
--> ceph-volume raw activate successful for osd ID: 0
[root@cv1-1 /]# CEPH_VOLUME_DEBUG=1 ceph-volume raw list /dev/foobar
{}
[root@cv1-1 /]#
```